Compile += / -= to bind & handle event (de)registration directly

vexx32 opened the pull request:
# PR Summary

Permits the following syntax:

```ps1
$scriptblock = { Write-Host 'event triggered' }
$ps = [powershell]::Create()
# Add event handler
$ps.InvocationStateChanged += $scriptblock

# Remove event handler
$ps.InvocationStateChanged -= $scriptblock
```

## Changes

- Expose `PSEvent` static and instance members of classes & enable tab completion for them.
- Add handling in the compiler to bind event (de)registration for `-=` or `+=` only if the LHS is a `MemberExpressionAst`, combined with a runtime check for the LHS value being `PSEvent`.
- Add a new binder type which will map the event member to its add/remove method as appropriate and then offload to `PSInvokeMemberBinder` with that method name.

## PR Context

See #12926 for some additional context, but my thoughts are essentially: events are currently not discoverable and very annoying to work with in PowerShell, sometimes requiring reliance on implementation details.

I do not believe this can be considered a breaking change, since the only place it applies is specifically for event members, which were previously pretty much entirely inaccessible prior to this change.

Resolves #12926 

With respect to tooling, there shouldn't be any breaking points, but event members will be offered in tab completion results with these changes, which may be considered unexpected in some circumstances.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [x] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 committed:
:sparkles: Let events be accessed by name

This gives visibility into events so that users can target them.
It also vastly simplifies the necessary handling in the compiler/binders
as we can use a runtime conditional expression to determine if the
target of the += / -= op is a PSEvent and maintain some sanity in the
binders and compiler logic needed here.

Completion logic also needed to be enhanced to provide completion for
event names.

vexx32 committed:
:sparkles: Add binder for event delegates

Adds a new binder to handle adding/removing event delegates with += / -=

If the LHS is a MemberAccessExpressionAst, the compiler will compile
+=/-= expressions into conditionals that first check if the LHS is a
PSEvent member. If so, it attempts to register the RHS argument as a
delegate for that event.

If the LHS isn't PSEvent, falls back to the typical compound assignment.

vexx32 committed:
:white_check_mark: Add eventing tests

Ensure we check both instance and static members to ensure we add or
remove the event correctly for both cases.

Also added a tab completion test for event members, which we don't seem
to have explicit tests for that I could find. As this completion is
related to this PR, seemed appropriate to add tests for that behaviour
here as well.

Co-authored-by: Ilya <darpa@yandex.ru>

vexx32 commented:
@TravisEz13 is it intentional that test results do not appear to be publicly viewable anywhere? Azure Pipelines reports a test pass percentage, but all the links lead to a blank page, and the Tests tab is missing from those pipeline jobs.

The few tests I could find failing in the log don't appear related at all, but I couldn't find all of them (and I'm not sure if there's a different way I need to be re-running failed jobs at the moment, too).

iSazonov commented:
@vexx32 There is a reference to test results
![image](https://user-images.githubusercontent.com/22290914/84228317-2b1ea600-ab00-11ea-8a7c-77c996ca5bc0.png)

vexx32 commented:
@iSazonov yep! But I can only see the log files, which lack any kind of summary as to exactly which failed. The logs are thousands of lines long, surely we're not expected to painfully read through all of them for a few tiny `-` characters and attempt to infer which tests might have failed. 🤔 

Oh, hold on... okay on one of the result pages I can see it, but the Azure Pipelines pages themselves don't seem to have a proper summary of results from what I can see.

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,104 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
-                case TokenKind.DivideEquals: et = ExpressionType.Divide; break;
-                case TokenKind.RemainderEquals: et = ExpressionType.Modulo; break;
-                case TokenKind.QuestionQuestionEquals: et = ExpressionType.Coalesce; break;
+                case TokenKind.Equals:
+                    return assignmentTarget.SetValue(this, right);
```
Since you refactor the code you could move the `Equals` in separate `if` block and convert the switch to a switch expression.

vexx32 commented:
@TravisEz13 filed here: https://developercommunity.visualstudio.com/content/problem/1072124/individual-test-results-failures-not-shown-in-buil.html

Didn't see any links pointing to a GH issues thing or anything, hopefully that's sufficient for now.

Yeah, that's about what I'm seeing. Makes it a bit tricky since our test output is very condensed and at most I just see rows of `+++++++` / `+++-+` marking individual tests passing/failing. Anyway, located the failing test for now... should have expected that one. 😁

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2101,6 +2101,69 @@ internal static BindingRestrictions GetVersionCheck(int expectedVersionNumber)
+    /// </summary>
+    internal class PSEventDelegateBinder : DynamicMetaObjectBinder
+    {
+        private static readonly PSEventDelegateBinder _addInstanceBinder = new PSEventDelegateBinder(
```
Please use `s_` prefix for statics. Below too.
```suggestion
        private static readonly PSEventDelegateBinder s_addInstanceBinder = new PSEventDelegateBinder(
```

vexx32 committed:
:recycle: Split code paths for assignment exprs

We can split the code path into assignments to members, and assignments
in other cases, avoiding the need to make some checks when the LHS is
not a MemberExpressionAst.

Reworked things a bit to use Lazy<string> for event definition.
Also refactored logic to use StringBuilder completely per Rob's comments

Co-authored-by: Robert Holt <rjmholt@gmail.com>

iSazonov commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -2664,6 +2664,33 @@ internal EventCacheEntry(EventInfo[] events)
+                        var sb = new StringBuilder();
+                        var invokeMethod = new MethodInformation(
+                            events[0].EventHandlerType.GetMethod("Invoke"), parametersToIgnore: 0).method;
+                        sb.AppendFormat(
+                            "event {0} {1}",
+                            ToStringCodeMethods.Type(events[0].EventHandlerType),
+                            events[0].Name);
+
+                        sb.Append("(");
+                        sb.AppendJoin(", ", invokeMethod.GetParameters().Select(p => $"{p.ParameterType.Name} {p.Name}"));
```
We think we should explicitly use InvariantCulture for formatting in the two places. It is our common pattern in Engine.

vexx32 replied:
`AppendJoin()` doesn't appear to have an overload that takes a format provider, but will do for AppendFormat().

SeeminglyScience replied:
I think @iSazonov is referring to the string interpolation used in the `Select` delegate.

iSazonov replied:
I meant both - AppendFormat and the string interpolation.

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,96 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
+                TokenKind.DivideEquals => ExpressionType.Divide,
+                TokenKind.RemainderEquals => ExpressionType.Modulo,
+                TokenKind.QuestionQuestionEquals => ExpressionType.Coalesce,
+                _ => throw new ArgumentException(nameof(tokenKind))
```
This changes original behavior.

vexx32 replied:
Yeah, you're right. I can just move the variable initialization down to here and have the default be `ExpressionType.Extension`, I missed that. Thanks!

iSazonov replied:
It seems original default is null.

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2101,6 +2101,69 @@ internal static BindingRestrictions GetVersionCheck(int expectedVersionNumber)
+            _isStatic = @static;
+        }
+
+        public override DynamicMetaObject Bind(DynamicMetaObject target, DynamicMetaObject[] args)
```
Maybe `internal`?

vexx32 replied:
I'll check, I think the binders in general have Bind() as a public member?

SeeminglyScience replied:
> I'll check, I think the binders in general have Bind() as a public member?

Yeah it's an abstract method.  Can't change accessibility modifier.

vexx32 replied:
Yep, `Bind()` is public. Think it has to be, since it's an `override` from the base class that's also public.

iSazonov replied:
#Closed.

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2101,6 +2101,69 @@ internal static BindingRestrictions GetVersionCheck(int expectedVersionNumber)
+            if (PSObject.Base(target.Value) is PSEvent @event)
+            {
+                string methodName = _isAddBinder
+                    ? $"add_{@event.Name}"
+                    : $"remove_{@event.Name}";
```
The same about InvariantCulture.

vexx32 commented:
MacOS Failure is a known issue and not a new failure: https://github.com/PowerShell/PowerShell/issues/12935

vexx32 commented:
@daxian-dbw friendly ping for review 😊

daxian-dbw commented:
> @daxian-dbw friendly ping for review 😊

Sorry for the delay ... reviewing this one now.

rjmholt commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -2664,6 +2664,41 @@ internal EventCacheEntry(EventInfo[] events)
+                        sb.Append(")");
+
+                        _eventDefinition = sb.ToString();
+                        sb.Clear();
```
This probably isn't needed -- it's probably simpler to let the GC just collect the whole string builder in one go when the method returns

rjmholt commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
+                return assignmentTarget.SetValue(this, right);
             }
 
+            var expressionType = tokenKind switch
```
```suggestion
            ExpressionType expressionType = tokenKind switch
```

rjmholt commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
+            return ReduceCompoundAssignment(this, assignmentTarget, lhsMemberAccessAst, expressionType, right);
+        }
+
+        internal static BlockExpression ReduceCompoundAssignment(
+            Compiler compiler,
```
Any reason to make this static, if we just have to pass `this` back in?

vexx32 replied:
Not anymore! I wrote it this way initially as I wasn't sure if I'd need to callback to this from the binder or some such witchcraft, but that's definitely not required here.

rjmholt commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
+                //      }
+                //
+                // The equivalent code is very similar for -= with the exception of calling GetRemoveMethod() instead of
+                // GetAddMethod().This lets us handle event delegates with += or -=, but we should only make this check
```
```suggestion
                // GetAddMethod(). This lets us handle event delegates with += or -=, but we should only make this check
```

rjmholt commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
-                exprs.Add(av.SetValue(this, Coalesce(getExpr, right)));
+                exprs.Add(assignmentTarget.SetValue(compiler, Coalesce(getExpr, valueToAssign)));
+            }
+            else if (lhsMemberExpressionAst != null
```
This suggests that this will only work when events are given in the form:

```powershell
$x.Event += $handler
```

But in the comment below you give an alternate form:

```powershell
$e = $x.Event
$e += $handler
```

Should we expect that to behave differently? If so, why? Is that consistent with other similar behaviours in PowerShell today?

vexx32 replied:
Hmm, that's a good point! I'm not sure, I guess I'd expect that that behaves the same. 🤔 

I was mostly mimicking C#'s syntax, not sure saving the event to a var works there, but I guess I'd expect that works also.

vexx32 replied:
Looking back at this, the example you're referring to is more illustrating how the `$object.EventName += $sb` expression is compiled (it's saved to a temporary variable first) rather than what would literally work.

Do we want to allow this to work for PSEvent instances saved in a variable as well? My concern here was mainly that the `$object.EventName += $b` is likely to be the most common usage, and I didn't want to bog down _every_ `$var += $sb` expression with a runtime check for `$var -is [PSEvent]` if I could avoid it.

vexx32 replied:
Looking at Sharplab, I can see the C# compiler actually completely prevents you assigning an event handler reference to a variable in the first place:
![image](https://user-images.githubusercontent.com/32407840/90328278-1bd93000-df69-11ea-91b2-a830f51dc00c.png)

I don't think we need to be _quite_ so strict in PS, but I think folks trying to use `+=` / `-=` for event handling will expect similar behaviour.

rjmholt replied:
Right, absolutely. I actually think you're right, but wanted to ask the question. (I referred to the compilation example because I knew it was a compilation example, but it *was* why I thought of it).

Events are strange second-class citizens in C# (I don't believe "event handler reference" is a meaningful concept for C#). But with that said, I'm not sure how many people we help by making them first-class in PowerShell (at the cost of things like runtime performance).

SeeminglyScience replied:
👍 for how it's currently implemented. It's consistent with C# and also with how `PSMethod`'s are handled.   Like you can `$method = $obj.ToString` but you can't then do `$method()`

rjmholt commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
+                ParameterExpression lhsStoreVar = Expression.Variable(typeof(object));
+                temps.Add(lhsStoreVar);
+
+                exprs.Add(Expression.Assign(lhsStoreVar, getExpr.Cast(typeof(object))));
```
What is the cast needed for here?

vexx32 replied:
I don't recall... I'll try it without, but I probably copied this pattern from elsewhere in the compiler.

rjmholt replied:
@daxian-dbw may know why it's needed

SeeminglyScience replied:
> What is the cast needed for here?

More than likely it's because expressions don't implicitly box, the cast here will do that.  Also matching expected types is super finicky (like a `Expression.Block`'s return type, or a variable) which casting everything to `object` greatly simplifies.

That's why there's a ton of casting, but I don't know off the top of my head of `getExpr` would already be cast. (Side note, it's also why byref/likes will probably never be supported 🙁)

rjmholt commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
+            ISupportsAssignment left,
+            TokenKind tokenKind,
+            Expression right,
+            MemberExpressionAst lhsMemberAccessAst)
```
Looking at how this is called, I think it does make sense the way it's currently written, but I just want to raise it for discussion.

A new possibility in one of the callers in this method now forces itself on all the other callers, leading to a number of calls where the caller must pass `null` in here. This seems to be common in the PowerShell codebase, but in my opinion leads to decohesion of methods, both from the perspective of the caller (which must now carefully select the appropriate subset of parameters) and from the perspective of the method itself (which must now carefully step between shared and scenario-specific logic).

In some respects this new parameter is a pseudo-boolean; it's either null or has a particular value that causes us to follow another path within the method.

That suggests to me that really we have two methods on our hands, with some shared logic best represented by breaking things up into smaller, composable methods.

vexx32 replied:
Definitely a good point. I can have a look at this and try to refactor things a bit.

vexx32 replied:
I think I have a pretty good way to do this. There's _some_ duplication of code, but it separates the code paths much more neatly. I'll commit this change shortly; any additional suggestions you have to improve what I managed to get to are more than welcome! 😊

rjmholt commented on src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:
```diff
@@ -249,7 +249,7 @@ internal void AddMembersByInferredTypesClrType(PSTypeName typename, bool isStati
                 var members = isStatic
                               ? PSObject.DotNetStaticAdapter.BaseGetMembers<PSMemberInfo>(type)
-                              : PSObject.DotNetInstanceAdapter.GetPropertiesAndMethods(type, false);
+                              : PSObject.DotNetInstanceAdapter.GetPropertiesMethodsAndEvents(type, false);
```
```suggestion
                              : PSObject.DotNetInstanceAdapter.GetPropertiesMethodsAndEvents(type, @static: false);
```

rjmholt commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -2663,6 +2663,45 @@ internal class EventCacheEntry : CacheEntry
+                    }
+                }
+
+                return sb.Append(")").ToString();
```
```suggestion
                return sb.Append(')').ToString();
```

vexx32 commented:
@rjmholt @iSazonov is there anything left for me to do here? 💖

iSazonov commented:
@vexx32 I am currently focusing on style fixes prior to the 7.1 release. I hope we have a couple of weeks for this.

vexx32 commented:
No worries, just wanted to make sure this didn't get forgotten about 💖

vexx32 commented:
@rjmholt @iSazonov  friendly ping for review 🙂

iSazonov commented:
@vexx32 I'm sorry, I'm out of the loop. The only thing I didn't see in the tests and what worries me is whether there is a problem with different Runspaces and threads?

vexx32 commented:
@iSazonov Yes and no. The majority of cases will "just work" with the already built in eventing logic, really. This functionally works just the same as you can go ahead and call `add_<event-name>($scriptblock)` on any object that has event handlers already, or `Register-ObjectEvent`.

Given we already have a cmdlet, this could be considered essentially syntax sugar (with some added discoverability) for similar functionality. I think there are some slight differences, but I'll be honest and say I'm not 100% sure what those will end up being.

My general understanding is that scriptblocks tend to be runspace-bound, and as a result will generally run on the thread that registered the event initially, when that thread is able to run them. @daxian-dbw or @rjmholt may be more familiar with the underpinnings here.

iSazonov commented:
@vexx32 Thanks for clarify!

> Expose event members as actual object members, allowing their names to register for tab completion, and the PSEvent metadata to be shown to users accessing the event.

Is this implemented? Have we tests for this?

vexx32 commented:
Aye, it is. There isn't an explicit test for it, come to think of it. The change to the tab completion test that comes back with `34` members instead of `33` now is a result of that, one of the members is an event.

I'll add an explicit one this afternoon, I think I can use a stopwatch / timer as a simple object with events on it that I can query for proving out the tab completion.

vexx32 commented:
@iSazonov I added the missing tab completion test. I also rebased to get this branch to pull in some of the newer changes and pass CI.

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -1005,10 +1005,17 @@ dir -Recurse `
             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Commands"
         }
 
+        It "completes event member names" {
```
Please start with capital letter:
```suggestion
        It "Completes event member names" {
```
Below too.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Eventing.Tests.ps1:
```diff
@@ -28,7 +30,7 @@ Describe "Event Subscriber Tests" -Tags "Feature" {
         Get-EventSubscriber | Should -BeNullOrEmpty
         Set-Variable incomingGlobal -Scope global -Value globVarValue
-        $null = Register-EngineEvent -SourceIdentifier foo -Action {Set-Variable -Scope global -Name aglobalvariable -Value $incomingGlobal}
+        $null = Register-EngineEvent -SourceIdentifier foo -Action { Set-Variable -Scope global -name aglobalvariable -Value $incomingGlobal }
```
```suggestion
        $null = Register-EngineEvent -SourceIdentifier foo -Action { Set-Variable -Scope global -Name aglobalvariable -Value $incomingGlobal }
```

iSazonov commented:
@daxian-dbw Could you please review or make a conclusion about closing the PR?

iSazonov commented:
/rebase

ALIENQuake commented:
Any news on this?

Shayan-To commented:
@vexx32 Is there any progress related to this? Is there any other issue/PR we can follow?

vexx32 commented:
I do not know if I personally have all that much time spare to look at this in detail but I will ping WG folks and to give it a look when they have a minute and get it on the radar, and put some labels on etc. It probably needs a rebase, at the least... I don't know if the WG folks will want to take it or not, but I think it's worth it myself.

(/cc @SeeminglyScience)

JamesWTruher commented:
it does look like a rebase is all that is needed (at least that's all i did, and the new tests were successful)

SeeminglyScience commented:
The Engine WG discussed this today (apologies for the wait). We worry that making this syntax easier (even if marginally) would imply a greater support for this operation than what we currently have (no closure created, handler must be invoked in a thread with a default runspace, potential runspace affinity issues, etc).

Thank you for your hard work on this Rain, and again sorry for the very late response.
